### PR TITLE
Slack integration

### PR DESCRIPTION
**Pull request by gstamp**
Adds a plugin for slack. This is very similar to the flowdock integration.

image

You will need to define `SLACK_TOKEN` in your environment to get this to work.

Resolves https://github.com/zendesk/samson/issues/52

### CODE REVIEW

File: plugins/slack/app/models/slack_notification.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  end
+
+  private
+
+  def content
+    subject = "[#{@project.name}] #{@deploy.summary}"
+    @content ||= SlackNotificationRenderer.render(@deploy, subject)
+  end
+
+  def url_helpers
```
Comment (sandlerr) [inline plugins/slack/app/models/slack_notification.rb]: this doesn't seem to be used?

File: plugins/slack/lib/samson_slack/samson_plugin.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
+require 'slack'
+
+module SamsonSlack
+  class Engine < Rails::Engine
+  end
+end
+
+Samson::Hooks.callback :stage_defined do
```
Comment (sandlerr) [inline plugins/slack/lib/samson_slack/samson_plugin.rb]: this hook is in flux at the moment, sorry about that: https://github.com/zendesk/samson/pull/324/files#diff-54c06d25726a01bbf1b82656eb8655b9R1

File: plugins/slack/lib/samson_slack/samson_plugin.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    def channel_exists?
+      if channel_name
+        errors.add(:slack_channels_name, "was not found") unless channel_for(channel_name)
+      end
+    end
+
+    def channel_for(name)
+      return nil unless name
+
+      response = Slack.channels_list(exlcude_archived: 1)
```
Comment (sandlerr) [inline plugins/slack/lib/samson_slack/samson_plugin.rb]: `exclude` typo

File: plugins/slack/lib/samson_slack/samson_plugin.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    def channel_name
+      slack_channels.first.try(:name)
+    end
+
+    def no_channel_name?(slack_attrs)
+      slack_attrs['name'].blank?
+    end
+
+    def update_channel_id
+      if channel_for(channel_name)
```
Comment (sandlerr) [inline plugins/slack/lib/samson_slack/samson_plugin.rb]: this makes an API call so it should probably be set to a local variable to reuse on the next line. Alternately you could memoise the return value of `channel_for`

Comment (gstamp) [inline plugins/slack/lib/samson_slack/samson_plugin.rb]: Yeah, I was thinking of wrapping this in a `Rails.cache.fetch` similar to the way you've wrapped some of your github calls.  Worth doing?

Comment (sandlerr) [inline plugins/slack/lib/samson_slack/samson_plugin.rb]: sounds good

File: plugins/slack/app/models/slack_notification.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  def deliver
+    Slack.chat_postMessage(
+      token: ENV['SLACK_TOKEN'],
+      channel: @stage.slack_channels.first.channel_id,
+      text: content,
+      parse: "full")
+
+  rescue Slack::Error
```
Comment (sandlerr) [inline plugins/slack/app/models/slack_notification.rb]: ideally this would log

**Comment by sandlerr** [comment]
Looks good! Travis-CI is having a bit of trouble - might help to check in the gems from vendor/cache that are added in this branch

**Comment by sandlerr** [comment]
cc @zendesk/samson

**Comment by gstamp** [comment]
So how would you like to proceed with the hook? Just hold off until the other PR is merged and fix it then?

**Comment by sandlerr** [comment]
@gstamp if you can cache the API call as we discussed we can merge this now and change it if the hook changes.

**Comment by sandlerr** [comment]
:+1:

File: .env.example [hunk 6]
```diff
@@ -28,3 +28,5 @@ GITHUB_TOKEN={fill me in}
 # PROJECT_CREATED_NOTIFY_ADDRESS=bobby-the-security-auditor@yourcompany.com
 
 # DEPLOY_GROUP_FEATURE={optional, set to 1 to enable Environments and DeployGroups}
+
+# SLACK_TOKEN={ required for the slack integration }
```
Comment (jwswj) [inline .env.example]: can we get a new line at end of file here?

**Comment by jwswj** [comment]
:+1: (apart from the new line)

**Comment by gstamp** [comment]
Newline added.  Cache call should be in there now.

Regards,

Glen

**Comment by sandlerr** [comment]
Thanks @gstamp!

**Pull request merged**